tuist/tuist
# Adding support for project additional files

Part of https://github.com/tuist/tuist/issues/202

### Short description 📝

Often projects may include additional files or folder references that don't need to be part of any build phase (e.g. design references, documentation etc...)


### Solution 📦

`Workspace.swift` manifest now supports `additionalFiles`, the same concept can be added to the `Project.swift` manifest with the same rules.

- `.glob(pattern: "Documentation/**")`: glob pattern to one or more files
- `.folderReference(path: "Designs")`: single path to a directory that is added as a folder reference

Example:

```swift
let project = Project(name: "App",
                      targets: [
                           // ...
                      ],
                      additionalFiles: [
                         "Dangerfile.swift"
                         "Documentation/**",
                         .folderReference(path: "Designs")
                     ])
```

<img width="240" alt="Screen Shot 2019-03-31 at 8 29 54 PM" src="https://user-images.githubusercontent.com/11914919/55293927-c774f500-53f3-11e9-8249-b20fa1829948.png">


### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [x] Update `Project` manifest to include `additionalFiles`
- [x] Update models to translate globs to paths
- [x] Update `ProjectFileElements` to include `additionalFiles`
- [x] Include fixture

### Test Plan ⚒

- Run `tuist generate` within `fixtures/ios_app_with_custom_workspace/App`
- Open the generated workspace
- Verify `Dangerfile.swift` is included in the project but isn't part of the resource phase
- Verify `Documentation` files are included (with the appropriate group structure that reflects the file system) and that the files aren't copied to resources
- Verify `Server` folder reference is included
